File: sdk/typescript/src/actions/objects.ts
import { jsonRequest, listParams, pathId, requestInit, withParams } from "../internal/utils.js";
import type {
  CreateObjectRequest,
  FavoriteState,
  ListParams,
  ListResponse,
  ObjectAttachment,
  ObjectAttachmentResponse,
  ObjectBacklinksParams,
  ObjectBacklinksResponse,
  ObjectListItem,
  ObjectListParams,
  ObjectResponse,
  ObjectVersion,
  ObjectVersionResponse,
  PinState,
  ReuseObjectAttachmentRequest,
  UpdateObjectRequest,
  UploadObjectAttachmentParams,
  UUID,
} from "../types.js";
import { setBoolean, setNumber, setString } from "./params.js";
import type { KivalClientBase, KivalResponseClient, WithSignal } from "./types.js";

/** Parameters for {@link listObjects}. */
export type ListObjectsParameters = WithSignal<
  ObjectListParams & {
    workspaceId: UUID;
  }
>;

/** Return type for {@link listObjects}. */
export type ListObjectsReturnType = ListResponse<ObjectListItem>;

/** Lists objects in a workspace. */
export function listObjects(
  client: KivalClientBase,
  parameters: ListObjectsParameters,
): Promise<ListObjectsReturnType> {
  const { workspaceId, ...options } = parameters;
  const params = listParams(options);
  params.set("status", options.status ?? "active");
  params.set("order", options.order ?? "created");
  setBoolean(params, "favorited", options.favorited);
  setBoolean(params, "pinned", options.pinned);
  return client.requestJson<ListObjectsReturnType>(
    withParams(`/workspaces/${pathId(workspaceId)}/objects`, params),
    requestInit({}, options.signal),
  );
}

/** Sets or clears a personal object favorite. */
export function setObjectFavorite(
  client: KivalClientBase,
  parameters: WithSignal<{ workspaceId: UUID; objectId: UUID; favorited: boolean }>,
): Promise<FavoriteState> {
  const { workspaceId, objectId, favorited, signal } = parameters;
  return client.requestJson<FavoriteState>(
    `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/favorite`,
    requestInit({ method: favorited ? "POST" : "DELETE" }, signal),
  );
}

/** Sets or clears a personal object pin. */
export function setObjectPin(
  client: KivalClientBase,
  parameters: WithSignal<{ workspaceId: UUID; objectId: UUID; pinned: boolean }>,
): Promise<PinState> {
  const { workspaceId, objectId, pinned, signal } = parameters;
  return client.requestJson<PinState>(
    `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/pin`,
    requestInit({ method: pinned ? "POST" : "DELETE" }, signal),
  );
}

/** Parameters for {@link getObject}. */
export type GetObjectParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
}>;

/** Return type for {@link getObject}. */
export type GetObjectReturnType = ObjectResponse;

/** Gets an object by ID. */
export function getObject(
  client: KivalClientBase,
  parameters: GetObjectParameters,
): Promise<GetObjectReturnType> {
  const { workspaceId, objectId, signal } = parameters;
  return client.requestJson<ObjectResponse>(
    `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}`,
    requestInit({}, signal),
  );
}

/** Parameters for {@link createObject}. */
export type CreateObjectParameters = WithSignal<{
  workspaceId: UUID;
  input: CreateObjectRequest;
}>;

/** Return type for {@link createObject}. */
export type CreateObjectReturnType = ObjectResponse;

/** Creates an object and its initial version. */
export function createObject(
  client: KivalClientBase,
  parameters: CreateObjectParameters,
): Promise<CreateObjectReturnType> {
  const { workspaceId, input, signal } = parameters;
  return client.requestJson<ObjectResponse>(
    `/workspaces/${pathId(workspaceId)}/objects`,
    jsonRequest("POST", input, signal),
  );
}

/** Parameters for {@link updateObject}. */
export type UpdateObjectParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  input: UpdateObjectRequest;
}>;

/** Return type for {@link updateObject}. */
export type UpdateObjectReturnType = ObjectResponse;

/** Updates an object, creating a new current version only when state changes. */
export function updateObject(
  client: KivalClientBase,
  parameters: UpdateObjectParameters,
): Promise<UpdateObjectReturnType> {
  const { workspaceId, objectId, input, signal } = parameters;
  return client.requestJson<ObjectResponse>(
    `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}`,
    jsonRequest("PATCH", input, signal),
  );
}

/** Parameters for {@link archiveObject}. */
export type ArchiveObjectParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
}>;

/** Return type for {@link archiveObject}. */
export type ArchiveObjectReturnType = ObjectResponse;

/** Archives an object. */
export function archiveObject(
  client: KivalClientBase,
  parameters: ArchiveObjectParameters,
): Promise<ArchiveObjectReturnType> {
  const { workspaceId, objectId, signal } = parameters;
  return client.requestJson<ObjectResponse>(
    `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/archive`,
    requestInit({ method: "POST" }, signal),
  );
}

/** Parameters for {@link unarchiveObject}. */
export type UnarchiveObjectParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
}>;

/** Return type for {@link unarchiveObject}. */
export type UnarchiveObjectReturnType = ObjectResponse;

/** Unarchives an object. */
export function unarchiveObject(
  client: KivalClientBase,
  parameters: UnarchiveObjectParameters,
): Promise<UnarchiveObjectReturnType> {
  const { workspaceId, objectId, signal } = parameters;
  return client.requestJson<ObjectResponse>(
    `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/unarchive`,
    requestInit({ method: "POST" }, signal),
  );
}

/** Parameters for {@link uploadObjectAttachment}. */
export type UploadObjectAttachmentParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  params: UploadObjectAttachmentParams;
  body: BodyInit;
}>;

/** Return type for {@link uploadObjectAttachment}. */
export type UploadObjectAttachmentReturnType = ObjectAttachment;

/**
 * Uploads a body supported by the runtime's Fetch API and creates an object attachment.
 *
 * Use a streaming `BodyInit` for large or otherwise non-buffered uploads.
 */
export function uploadObjectAttachment(
  client: KivalClientBase,
  parameters: UploadObjectAttachmentParameters,
): Promise<UploadObjectAttachmentReturnType> {
  const { workspaceId, objectId, params, body, signal } = parameters;
  const searchParams = new URLSearchParams();
  setString(searchParams, "version_id", params.version_id);
  setString(searchParams, "name", params.name);
  setString(searchParams, "media_type", params.media_type);
  setString(searchParams, "metadata", params.metadata);

  return client
    .requestJson<ObjectAttachmentResponse>(
      withParams(
        `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/attachments/upload`,
        searchParams,
      ),
      requestInit(
        {
          method: "POST",
          ...(params.media_type ? { headers: { "content-type": params.media_type } } : {}),
          body,
        },
        signal,
      ),
    )
    .then((response) => response.attachment);
}

/** Parameters for {@link listObjectAttachments}. */
export type ListObjectAttachmentsParameters = WithSignal<
  ListParams & {
    workspaceId: UUID;
    objectId: UUID;
  }
>;

/** Return type for {@link listObjectAttachments}. */
export type ListObjectAttachmentsReturnType = ListResponse<ObjectAttachment>;

/** Lists object attachments. */
export function listObjectAttachments(
  client: KivalClientBase,
  parameters: ListObjectAttachmentsParameters,
): Promise<ListObjectAttachmentsReturnType> {
  const { workspaceId, objectId, ...options } = parameters;
  return client.requestJson<ListObjectAttachmentsReturnType>(
    withParams(
      `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/attachments`,
      listParams(options),
    ),
    requestInit({}, options.signal),
  );
}

/** Parameters for {@link getObjectAttachment}. */
export type GetObjectAttachmentParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  attachmentId: UUID;
}>;

/** Return type for {@link getObjectAttachment}. */
export type GetObjectAttachmentReturnType = ObjectAttachment;

/** Gets object-attachment metadata by ID. */
export function getObjectAttachment(
  client: KivalClientBase,
  parameters: GetObjectAttachmentParameters,
): Promise<GetObjectAttachmentReturnType> {
  const { workspaceId, objectId, attachmentId, signal } = parameters;
  return client
    .requestJson<ObjectAttachmentResponse>(
      `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}` +
        `/attachments/${pathId(attachmentId)}`,
      requestInit({}, signal),
    )
    .then((response) => response.attachment);
}

/** Parameters for {@link reuseObjectAttachment}. */
export type ReuseObjectAttachmentParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  input: ReuseObjectAttachmentRequest;
}>;

/** Return type for {@link reuseObjectAttachment}. */
export type ReuseObjectAttachmentReturnType = ObjectAttachment;

/** Creates an attachment by reusing an authorized source attachment. */
export function reuseObjectAttachment(
  client: KivalClientBase,
  parameters: ReuseObjectAttachmentParameters,
): Promise<ReuseObjectAttachmentReturnType> {
  const { workspaceId, objectId, input, signal } = parameters;
  return client
    .requestJson<ObjectAttachmentResponse>(
      `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/attachments/reuse`,
      jsonRequest("POST", input, signal),
    )
    .then((response) => response.attachment);
}

/** Parameters for {@link getObjectAttachmentContent}. */
export type GetObjectAttachmentContentParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  attachmentId: UUID;
}>;

/** Return type for {@link getObjectAttachmentContent}. */
export type GetObjectAttachmentContentReturnType = Uint8Array;

/**
 * Fetches object-attachment content as bytes.
 *
 * This convenience action buffers the complete response. Use
 * {@link getObjectAttachmentContentResponse} to consume large bodies incrementally.
 */
export function getObjectAttachmentContent(
  client: KivalResponseClient,
  parameters: GetObjectAttachmentContentParameters,
): Promise<GetObjectAttachmentContentReturnType> {
  const { workspaceId, objectId, attachmentId, signal } = parameters;
  return client.requestBytes(
    `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}` +
      `/attachments/${pathId(attachmentId)}/content`,
    requestInit({}, signal),
  );
}

/** Parameters for {@link getObjectAttachmentContentResponse}. */
export type GetObjectAttachmentContentResponseParameters = GetObjectAttachmentContentParameters;

/** Return type for {@link getObjectAttachmentContentResponse}. */
export type GetObjectAttachmentContentResponseReturnType = Response;

/**
 * Fetches an object-attachment response without buffering its body.
 *
 * The transport timeout covers receiving the response headers. Body-stream errors after this
 * action resolves are reported by the runtime's Fetch API.
 */
export function getObjectAttachmentContentResponse(
  client: KivalResponseClient,
  parameters: GetObjectAttachmentContentResponseParameters,
): Promise<GetObjectAttachmentContentResponseReturnType> {
  const { workspaceId, objectId, attachmentId, signal } = parameters;
  return client.requestResponse(
    `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}` +
      `/attachments/${pathId(attachmentId)}/content`,
    requestInit({}, signal),
  );
}

/** Parameters for {@link listObjectVersions}. */
export type ListObjectVersionsParameters = WithSignal<
  ListParams & {
    workspaceId: UUID;
    objectId: UUID;
  }
>;

/** Return type for {@link listObjectVersions}. */
export type ListObjectVersionsReturnType = ListResponse<ObjectVersion>;

/** Lists object versions. */
export function listObjectVersions(
  client: KivalClientBase,
  parameters: ListObjectVersionsParameters,
): Promise<ListObjectVersionsReturnType> {
  const { workspaceId, objectId, ...options } = parameters;
  return client.requestJson<ListObjectVersionsReturnType>(
    withParams(
      `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/versions`,
      listParams(options),
    ),
    requestInit({}, options.signal),
  );
}

/** Parameters for {@link getObjectVersion}. */
export type GetObjectVersionParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  version: UUID | number;
}>;

/** Return type for {@link getObjectVersion}. */
export type GetObjectVersionReturnType = ObjectVersion;

/** Gets an object version by immutable ID or monotonic version number. */
export function getObjectVersion(
  client: KivalClientBase,
  parameters: GetObjectVersionParameters,
): Promise<GetObjectVersionReturnType> {
  const { workspaceId, objectId, version, signal } = parameters;
  if (typeof version === "number" && (!Number.isSafeInteger(version) || version < 1)) {
    throw new TypeError("version must be a positive safe integer");
  }

  return client
    .requestJson<ObjectVersionResponse>(
      `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/versions/${pathId(version.toString())}`,
      requestInit({}, signal),
    )
    .then((response) => response.version);
}

/** Parameters for {@link getObjectBacklinks}. */
export type GetObjectBacklinksParameters = WithSignal<
  ObjectBacklinksParams & {
    workspaceId: UUID;
    objectId: UUID;
  }
>;

/** Return type for {@link getObjectBacklinks}. */
export type GetObjectBacklinksReturnType = ObjectBacklinksResponse;

/** Lists visible inbound explicit edges and textual references for an object. */
export function getObjectBacklinks(
  client: KivalClientBase,
  parameters: GetObjectBacklinksParameters,
): Promise<GetObjectBacklinksReturnType> {
  const { workspaceId, objectId, ...options } = parameters;
  const params = new URLSearchParams();
  setNumber(params, "limit", options.limit);
  setString(params, "edge_cursor", options.edge_cursor);
  setString(params, "reference_cursor", options.reference_cursor);
  if (options.include_archived) params.set("include_archived", "true");
  return client.requestJson<ObjectBacklinksResponse>(
    withParams(`/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/backlinks`, params),
    requestInit({}, options.signal),
  );
}

File: sdk/rust/src/client/objects.rs
//! Object and object-version API operations.

use reqwest::{Body, Method, Response};
use uuid::Uuid;

use crate::{
    ClientError, CreateObjectRequest, KivalClient, ListParams, ListResponse, ObjectAttachment,
    ObjectAttachmentResponse, ObjectBacklinksParams, ObjectBacklinksResponse, ObjectListItem,
    ObjectListParams, ObjectResponse, ObjectVersion, ObjectVersionResponse,
    ReuseObjectAttachmentRequest, Transport, UpdateObjectRequest, UploadObjectAttachmentParams,
    client::transport::append_list_params,
};

/// Identifier accepted by the object-version read endpoint.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ObjectVersionIdentifier {
    /// Immutable object-version UUID.
    Id(Uuid),
    /// Monotonic object version number.
    Number(i64),
}

impl From<Uuid> for ObjectVersionIdentifier {
    fn from(value: Uuid) -> Self {
        Self::Id(value)
    }
}

impl From<i64> for ObjectVersionIdentifier {
    fn from(value: i64) -> Self {
        Self::Number(value)
    }
}

impl std::fmt::Display for ObjectVersionIdentifier {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Id(version_id) => version_id.fmt(formatter),
            Self::Number(version_number) => version_number.fmt(formatter),
        }
    }
}

impl<S> KivalClient<S>
where
    S: Transport,
{
    /// Lists objects in a workspace.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn list_objects(
        &self,
        workspace_id: Uuid,
        params: &ObjectListParams,
    ) -> Result<ListResponse<ObjectListItem>, ClientError> {
        self.require_api_key()?;

        let mut url = self.api_url(&format!("/workspaces/{workspace_id}/objects"))?;
        append_list_params(
            &mut url,
            &ListParams { limit: params.limit, cursor: params.cursor.clone() },
        );
        {
            let mut pairs = url.query_pairs_mut();
            pairs
                .append_pair("status", params.status.as_str())
                .append_pair("order", params.order.as_str());

            if let Some(favorited) = params.favorited {
                pairs.append_pair("favorited", if favorited { "true" } else { "false" });
            }

            if let Some(pinned) = params.pinned {
                pairs.append_pair("pinned", if pinned { "true" } else { "false" });
            }
        }

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Creates an object and its initial version.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn create_object(
        &self,
        workspace_id: Uuid,
        request: CreateObjectRequest,
    ) -> Result<ObjectResponse, ClientError> {
        self.require_api_key()?;

        let request = self
            .request(&Method::POST, &format!("/workspaces/{workspace_id}/objects"))?
            .json(&request);

        self.send_json(request).await
    }

    /// Gets an object by ID.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn get_object(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
    ) -> Result<ObjectResponse, ClientError> {
        self.require_api_key()?;

        let request =
            self.request(&Method::GET, &format!("/workspaces/{workspace_id}/objects/{object_id}"))?;

        self.send_json(request).await
    }

    /// Updates an object, creating a new current version only when state changes.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn update_object(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        request: UpdateObjectRequest,
    ) -> Result<ObjectResponse, ClientError> {
        self.require_api_key()?;

        let request = self
            .request(&Method::PATCH, &format!("/workspaces/{workspace_id}/objects/{object_id}"))?
            .json(&request);

        self.send_json(request).await
    }

    /// Archives an object.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn archive_object(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
    ) -> Result<ObjectResponse, ClientError> {
        self.require_api_key()?;

        let request = self.request(
            &Method::POST,
            &format!("/workspaces/{workspace_id}/objects/{object_id}/archive"),
        )?;

        self.send_json(request).await
    }

    /// Unarchives an object.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn unarchive_object(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
    ) -> Result<ObjectResponse, ClientError> {
        self.require_api_key()?;

        let request = self.request(
            &Method::POST,
            &format!("/workspaces/{workspace_id}/objects/{object_id}/unarchive"),
        )?;

        self.send_json(request).await
    }

    /// Lists visible inbound explicit edges and textual references for an object.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn get_object_backlinks(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        params: &ObjectBacklinksParams,
    ) -> Result<ObjectBacklinksResponse, ClientError> {
        self.require_api_key()?;

        let mut url =
            self.api_url(&format!("/workspaces/{workspace_id}/objects/{object_id}/backlinks"))?;
        {
            let mut pairs = url.query_pairs_mut();
            if let Some(limit) = params.limit {
                pairs.append_pair("limit", &limit.to_string());
            }
            if let Some(edge_cursor) = &params.edge_cursor {
                pairs.append_pair("edge_cursor", edge_cursor);
            }
            if let Some(reference_cursor) = &params.reference_cursor {
                pairs.append_pair("reference_cursor", reference_cursor);
            }
            if params.include_archived {
                pairs.append_pair("include_archived", "true");
            }
        }

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Lists object attachments.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn list_object_attachments(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        params: &ListParams,
    ) -> Result<ListResponse<ObjectAttachment>, ClientError> {
        self.require_api_key()?;

        let mut url =
            self.api_url(&format!("/workspaces/{workspace_id}/objects/{object_id}/attachments"))?;
        append_list_params(&mut url, params);

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Uploads in-memory bytes and creates an object attachment.
    ///
    /// Use [`Self::upload_object_attachment_body`] for streaming or otherwise non-buffered bodies.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn upload_object_attachment(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        params: &UploadObjectAttachmentParams,
        bytes: Vec<u8>,
    ) -> Result<ObjectAttachment, ClientError> {
        self.upload_object_attachment_body(workspace_id, object_id, params, Body::from(bytes)).await
    }

    /// Uploads a Reqwest body and creates an object attachment.
    ///
    /// A streaming body can be created with [`Body::wrap_stream`].
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn upload_object_attachment_body(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        params: &UploadObjectAttachmentParams,
        body: Body,
    ) -> Result<ObjectAttachment, ClientError> {
        self.require_api_key()?;

        let mut url = self.api_url(&format!(
            "/workspaces/{workspace_id}/objects/{object_id}/attachments/upload"
        ))?;
        {
            let mut pairs = url.query_pairs_mut();
            if let Some(version_id) = params.version_id {
                pairs.append_pair("version_id", &version_id.to_string());
            }
            if let Some(name) = &params.name {
                pairs.append_pair("name", name);
            }
            if let Some(media_type) = &params.media_type {
                pairs.append_pair("media_type", media_type);
            }
            if let Some(metadata) = &params.metadata {
                pairs.append_pair("metadata", metadata);
            }
        }

        let request = self.request_url(&Method::POST, url)?.body(body);
        let response = self.send_json::<ObjectAttachmentResponse>(request).await?;
        Ok(response.attachment)
    }

    /// Creates an attachment by reusing an authorized source attachment.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn reuse_object_attachment(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        request: ReuseObjectAttachmentRequest,
    ) -> Result<ObjectAttachment, ClientError> {
        self.require_api_key()?;

        let request = self
            .request(
                &Method::POST,
                &format!("/workspaces/{workspace_id}/objects/{object_id}/attachments/reuse"),
            )?
            .json(&request);
        let response = self.send_json::<ObjectAttachmentResponse>(request).await?;

        Ok(response.attachment)
    }

    /// Fetches object attachment content bytes by ID.
    ///
    /// This convenience method buffers the complete response. Use
    /// [`Self::get_object_attachment_content_response`] to consume large bodies incrementally.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn get_object_attachment_content(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        attachment_id: Uuid,
    ) -> Result<Vec<u8>, ClientError> {
        let response = self
            .get_object_attachment_content_response(workspace_id, object_id, attachment_id)
            .await?;
        Ok(response.bytes().await?.to_vec())
    }

    /// Fetches an object attachment response without buffering its body.
    ///
    /// Callers can consume the response incrementally with [`Response::chunk`] or
    /// [`Response::bytes_stream`].
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn get_object_attachment_content_response(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        attachment_id: Uuid,
    ) -> Result<Response, ClientError> {
        self.require_api_key()?;

        let request = self.request(
            &Method::GET,
            &format!(
                "/workspaces/{workspace_id}/objects/{object_id}/attachments/{attachment_id}/content"
            ),
        )?;
        self.send(request).await
    }

    /// Gets an object attachment by ID.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn get_object_attachment(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        attachment_id: Uuid,
    ) -> Result<ObjectAttachment, ClientError> {
        self.require_api_key()?;

        let request = self.request(
            &Method::GET,
            &format!("/workspaces/{workspace_id}/objects/{object_id}/attachments/{attachment_id}"),
        )?;
        let response = self.send_json::<ObjectAttachmentResponse>(request).await?;

        Ok(response.attachment)
    }

    /// Lists object versions.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn list_object_versions(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        params: &ListParams,
    ) -> Result<ListResponse<ObjectVersion>, ClientError> {
        self.require_api_key()?;

        let mut url =
            self.api_url(&format!("/workspaces/{workspace_id}/objects/{object_id}/versions"))?;
        append_list_params(&mut url, params);

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Gets an object version by immutable ID or monotonic version number.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured. Numeric identifiers
    /// smaller than one are rejected by the server.
    pub async fn get_object_version(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        version: impl Into<ObjectVersionIdentifier>,
    ) -> Result<ObjectVersion, ClientError> {
        self.require_api_key()?;

        let version = version.into();
        let request = self.request(
            &Method::GET,
            &format!("/workspaces/{workspace_id}/objects/{object_id}/versions/{version}"),
        )?;
        let response = self.send_json::<ObjectVersionResponse>(request).await?;

        Ok(response.version)
    }
}
